File: macruby-appscript/trunk/lib/_appscript/referencerenderer.rb
# Copyright (C) 2008 HAS. 
# Released under MIT License.

framework 'Appscript'


class MRAReferenceRenderer < AEMResolver
	# Generates string representations of appscript references from aem object specifiers.

	private_class_method :new
	attr_reader :result
	
	def initWithAppData(app_data)
		init
		@_app_data = app_data
		@result = ""
		return self
	end
	
	def _format(val)
		case val
			when AEMQuery
				return MRAReferenceRenderer.render(@_app_data, val)
			when NSString
				return val.to_s.inspect
			# TO DO: reformat ObjC values as Ruby values
			when NSObject
				return val
		else
			return val.inspect
		end
	end
	
	##
	
	def property(code)
		code = AEMType.typeWithCode(code)
		name = @_app_data.property_by_code.objectForKey(code)
		name = @_app_data.element_by_code.objectForKey(code) if not name
		name = "<#{code}>" if not name
		@result += ".#{name}"
		return self
	end
	
	def elements(code)
		code = AEMType.typeWithCode(code)
		name = @_app_data.element_by_code.objectForKey(code)
		name = @_app_data.property_by_code.objectForKey(code) if not name
		name = "<#{code}>" if not name
		@result += ".#{name}"
		return self
	end
	
	def byName(name)
		@result += "[#{_format(name)}]"
		return self
	end
	
	def byIndex(index)
		@result += "[#{_format(index)}]"
		return self
	end
	
	def byID(id)
		@result += ".ID(#{_format(id)})"
		return self
	end
	
	def byRange_to(sel1, sel2)
		@result += "[#{_format(sel1)}, #{_format(sel2)}]"
		return self
	end
	
	def byTest(sel)
		@result += "[#{_format(sel)}]"
		return self
	end
	
	def previous(sel)
		@result += ".previous(#{_format(@_app_data.type_by_code[sel])})"
		return self
	end
	
	def next(sel)
		@result += ".next(#{_format(@_app_data.type_by_code[sel])})"
		return self
	end
	
	##
	
	def customRoot(value)
		app
		@result += ".AS_new_reference(#{_format(value)})"
		return self
	end
	
	ConstructorNames = {
			KASTargetCurrent => :current,
			KASTargetName => :by_name,
			KASTargetBundleID => :by_id,
			KASTargetURL => :by_url,
			KASTargetPID => :by_pid,
			KASTargetDescriptor => :by_aedesc, # TO DO
	}
	
	def app
		case @_app_data.constructor
			when KASTargetCurrent
				@result = "app.current"
			when KASTargetName
				@result = "app(#{_format(@_app_data.identifier)})"
		else
			@result = "app.#{ConstructorNames[@_app_data.constructor]}(#{_format(@_app_data.identifier)})"
		end
		return self
	end
	
	def con
		@result = "con"
		return self
	end
	
	def its
		@result = "its"
		return self
	end
	
	##
	
	def beginning
		@result += ".beginning"
		return self
	end
	
	def end
		@result += ".end"
		return self
	end
	
	def before
		@result += ".before"
		return self
	end
	
	def after
		@result += ".after"
		return self
	end
	
	##
	
	def first
		@result += ".first"
		return self
	end
	
	def middle
		@result += ".middle"
		return self
	end
	
	def last
		@result += ".last"
		return self
	end
	
	def any
		@result += ".any"
		return self
	end
	
	##
	
	def greaterThan(val)
		@result += ".gt(#{_format(val)})"
		return self
	end
	
	def greaterOrEquals(val)
		@result += ".ge(#{_format(val)})"
		return self
	end
	
	def equals(val)
		@result += ".eq(#{_format(val)})"
		return self
	end
	
	def notEquals(val)
		@result += ".ne(#{_format(val)})"
		return self
	end
	
	def lessThan(val)
		@result += ".lt(#{_format(val)})"
		return self
	end
	
	def lessOrEquals(val)
		@result += ".le(#{_format(val)})"
		return self
	end
	
	def beginsWith(val)
		@result += ".begins_with(#{_format(val)})"
		return self
	end
	
	def endsWith(val)
		@result += ".ends_with(#{_format(val)})"
		return self
	end
	
	def contains(val)
		@result += ".contains(#{_format(val)})"
		return self
	end
	
	def isIn(val)
		@result += ".is_in(#{_format(val)})"
		return self
	end
	
	def AND(operands)
		@result += ".and(#{(operands.map { |val| _format(val) }).join(', ')})" # TO DO: check
		return self
	end
	
	def OR(operands)
		@result += ".or(#{(operands.map { |val| _format(val) }).join(', ')})"
		return self
	end
	
	def NOT
		@result += ".not"
		return self
	end
	
	# public
	
	def MRAReferenceRenderer.render(app_data, aem_reference)
		f = self.alloc.initWithAppData(app_data)
		begin
			aem_reference.resolveWithObject(f)
			return f.result
		rescue
			return "#{new(app_data).app.result}.AS_new_reference(#{_format(aem_reference)})"
		end
	end

end

	
	

File: objc-appscript/tags/objc-appscript-0.5.1/src/Appscript/referencerenderer.m
//
//  referencerenderer.m
//  appscript
//

#import "referencerenderer.h"


/**********************************************************************/
// reference renderer base

@implementation ASReferenceRenderer

// clients should avoid calling this method directly; use +formatObject:appData: instead
- (id)initWithAppData:(id)appData_ {
	self = [super init];
	if (!self) return self;
	result = [NSMutableString string];
	appData = [appData_ retain];
	return self;
}

- (void)dealloc {
	[appData release];
	[super dealloc];
}



/*******/

// takes an AEM specifier/test object, ASAppData instance; returns string representation
+ (NSString *)formatObject:(id)object appData:(id)appData_ {
	ASReferenceRenderer *renderer;
	NSString *string;
	
	if ([object isKindOfClass: [AEMQuery class]]) {
		renderer = [[[self class] alloc] initWithAppData: appData_];
		[object resolveWithObject: renderer];
		string = [renderer result];
		if (!string)
			string = [NSString stringWithFormat: @"[%@ AS_referenceWithObject: %@]", [renderer app], object];
		[renderer release];
	} else {
		string = [AEMObjectRenderer formatObject: object];
	}
	return string;
}

- (NSString *)result {
	return result;
}

/*******/

// method stubs; application-specific subclasses should override to provide code->name translations
- (NSString *)propertyByCode:(OSType)code { 
	return nil;
}

- (NSString *)elementByCode:(OSType)code {
	return nil;
}

// method stub; application-specific subclasses should override to provide class name prefix

- (NSString *)prefix {
	return @"AS";
}

/*******/

- (NSString *)format:(id)object {
	if ([object isKindOfClass: [AEMQuery class]])
		return [[self class] formatObject: object appData: appData];
	else
		return [AEMObjectRenderer formatObject: object];
}

/*******/

- (ASReferenceRenderer *)property:(OSType)code {
	NSString *name;
	
	name = [self propertyByCode: code];
	if (!name)
		name = [self elementByCode: code];
	if (!name) { // no code->name translation available
		result = nil;
		return nil;
	}
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" %@]", name];
    return self;
}

- (ASReferenceRenderer *)elements:(OSType)code {
	NSString *name;
	
	name = [self elementByCode: code];
	if (!name)
		name = [self propertyByCode: code];
	if (!name) { // no code->name translation available
		result = nil;
		return nil;
	}
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" %@]", name];
    return self;
}


- (ASReferenceRenderer *)first {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" first]"];
    return self;
}

- (ASReferenceRenderer *)middle {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" middle]"];
    return self;
}

- (ASReferenceRenderer *)last {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" last]"];
    return self;
}

- (ASReferenceRenderer *)any {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" any]"];
    return self;
}

// by-index, by-name, by-id selectors

- (ASReferenceRenderer *)byIndex:(id)index {
	[result insertString: @"[" atIndex: 0];
	if ([index isKindOfClass: [NSNumber class]])
		[result appendFormat: @" at: %@]", index];
	else
		[result appendFormat: @" byIndex: %@]", [self format: index]];
    return self;
}

- (ASReferenceRenderer *)byName:(id)name {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" byName: %@]", [self format: name]];
    return self;
}

- (ASReferenceRenderer *)byID:(id)id_ {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" byID: %@]", [self format: id_]];
    return self;
}

// by-relative-position selectors

- (ASReferenceRenderer *)previous:(OSType)class_ {
	NSAppleEventDescriptor *desc = [NSAppleEventDescriptor descriptorWithTypeCode:class_];
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" previous: %@]", [appData unpack: desc]];
    return self;
}

- (ASReferenceRenderer *)next:(OSType)class_ {
	NSAppleEventDescriptor *desc = [NSAppleEventDescriptor descriptorWithTypeCode:class_];
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" next: %@]", [appData unpack: desc]];
    return self;
}

// by-range selector

- (ASReferenceRenderer *)byRange:(id)fromObject to:(id)toObject {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" byRange: %@ to: %@]",
						  [self format: fromObject],
						  [self format: toObject]];
    return self;
}

// by-test selector

- (ASReferenceRenderer *)byTest:(id)testReference {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" byTest: %@]", [self format: testReference]];
    return self;
}

// insertion location selectors

- (ASReferenceRenderer *)beginning {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" beginning]"];
    return self;
}

- (ASReferenceRenderer *)end {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" end]"];
    return self;
}

- (ASReferenceRenderer *)before {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" before]"];
    return self;
}

- (ASReferenceRenderer *)after {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" after]"];
    return self;
}

// test clause renderers

- (ASReferenceRenderer *)greaterThan:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" greaterThan: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)greaterOrEquals:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" greaterOrEquals: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)equals:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" equals: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)notEquals:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" notEquals: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)lessThan:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" lessThan: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)lessOrEquals:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" lessOrEquals: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)beginsWith:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" beginsWith: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)endsWith:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" endsWith: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)contains:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" contains: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)isIn:(id)object {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" isIn: %@]", [self format: object]];
    return self;
}

- (ASReferenceRenderer *)AND:(id)remainingOperands {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" AND: %@]", [self format: remainingOperands]];
    return self;
}

- (ASReferenceRenderer *)OR:(id)remainingOperands {
	[result insertString: @"[" atIndex: 0];
	[result appendFormat: @" OR: %@]", [self format: remainingOperands]];
    return self;
}

- (ASReferenceRenderer *)NOT {
	[result insertString: @"[" atIndex: 0];
	[result appendString: @" NOT]"];
    return self;
}


// reference roots

- (ASReferenceRenderer *)app {
	if (appData) {
		NSError *error;
		id target = [appData targetWithError: &error];
		if (target) {
			AEMTargetType targetType = [target targetType];
			id targetData = [target targetData];
			switch (targetType) {
				case kAEMTargetCurrent:
					[result appendFormat: @"[%@Application application]", [self prefix]];
					return self;
				case kAEMTargetFileURL:
					[result appendFormat: @"[%@Application applicationWithName: %@]", 
											[self prefix], [self format: [targetData path]]];
					return self;
				case kAEMTargetEppcURL:
					[result appendFormat: @"[%@Application applicationWithURL: %@]", 
											[self prefix], [self format: targetData]];
					return self;
				case kAEMTargetPID:
					[result appendFormat: @"[%@Application applicationWithPID: %@]", 
											[self prefix], [self format: targetData]];
					return self;
				case kAEMTargetDescriptor:
					[result appendFormat: @"[%@Application applicationWithDescriptor: %@]", 
											[self prefix], [self format: targetData]];
					return self;
			}
		}
		[result appendFormat: @"<%@Application invalid target (error=%i)>", [self prefix], [error code]];
		return self;
	}
	[result appendFormat: @"%@App", [self prefix]];
	return self;
}

- (ASReferenceRenderer *)con {
	[result appendFormat: @"%@Con", [self prefix]];
    return self;
}

- (ASReferenceRenderer *)its {
	[result appendFormat: @"%@Its", [self prefix]];
    return self;
}

- (ASReferenceRenderer *)customRoot:(id)rootObject {
	[result appendFormat: @"%@CustomRoot(%@)", [self prefix], rootObject];
    return self;
}

@end
